objects/src/accounts/vault.rs

```diff
@@ -17,37 +20,41 @@ use core::default::Default;
 /// An account vault can be reduced to a single hash which is the root of the Sparse Merkle tree.
 #[derive(Debug, Clone, Default, PartialEq, Eq)]
 pub struct AccountVault {
-    // TODO: add backing sparse Merkle tree
-    assets: Vec<Asset>,
+    asset_tree: TieredSmt,
 }
 
 impl AccountVault {
     // CONSTRUCTOR
     // --------------------------------------------------------------------------------------------
     /// Returns a new account vault initialized with the provided assets.
-    ///
-    /// TODO: return error if there are duplicates in the provided asset list.
-    pub fn new(assets: &[Asset]) -> Self {
-        Self {
-            // TODO: put assets into a Sparse Merkle trees
-            assets: assets.to_vec(),
-        }
+    pub fn new(assets: &[Asset]) -> Result<Self, AccountError> {
+        Ok(Self {
+            asset_tree: TieredSmt::with_leaves(
+                assets.iter().map(|asset| (asset.key().into(), (*asset).into())),
+            )
+            .map_err(AccountError::DuplicateAsset)?,
+        })
     }
 
     // PUBLIC ACCESSORS
     // --------------------------------------------------------------------------------------------
 
     /// Returns a commitment to this vault.
-    pub fn root(&self) -> Digest {
-        Digest::default()
+    pub fn commitment(&self) -> Digest {
+        self.asset_tree.root()
     }
 
     /// Returns true if the specified non-fungible asset is stored in this vault.
     pub fn has_non_fungible_asset(&self, asset: Asset) -> Result<bool, AccountError> {
         if asset.is_fungible() {
             return Err(AccountError::not_a_non_fungible_asset(asset));
         }
-        todo!()
+
+        // check if the asset is stored in the vault
+        match self.asset_tree.get_value(asset.key().into()) {
+            asset if asset == [ZERO, ZERO, ZERO, ZERO] => Ok(false),
```

Comment:
Instead of using `[ZERO, ZERO, ZERO, ZERO]` could we define something like an `EMPTY_WORD` constant and use it here (and in other places)? Also, I wonder if this constant should actually be defined in the `TieredSmt` struct.

Reply:
Yes I agree.  I like the latter proposal of having something like `EMPTY_WORD` or `EMPTY_VALUE` as an associated constant in `TieredSmt`.  

I have gone with the former proposal for now of using the `EMPTY_WORD` constant defined in the `vault` module and added a `TODO` for once the associated constant is added to `TieredSmt`.

Reply:
Issue created to track this - https://github.com/0xPolygonMiden/miden-base/issues/146.